libeos-payg/service.c

```diff
@@ -125,22 +126,90 @@ epg_service_get_main_option_entries (GssService *service)
   return g_steal_pointer (&entries);
 }
 
+static void provider_load_cb (GObject      *source_object,
+                              GAsyncResult *result,
+                              gpointer      user_data);
 static void manager_new_cb (GObject      *source_object,
                             GAsyncResult *result,
                             gpointer      user_data);
+static void epg_service_startup_complete (EpgService  *self,
+                                          GTask       *task,
+                                          EpgProvider *provider);
 
 static void
 epg_service_startup_async (GssService          *service,
                            GCancellable        *cancellable,
                            GAsyncReadyCallback  callback,
                            gpointer             user_data)
 {
-  EpgService *self = EPG_SERVICE (service);
   g_autoptr(GError) local_error = NULL;
 
   g_autoptr(GTask) task = g_task_new (service, cancellable, callback, user_data);
   g_task_set_source_tag (task, epg_service_startup_async);
 
+  g_autoptr(EpgProviderLoader) loader = epg_provider_loader_new (NULL);
+  epg_provider_loader_load_async (loader, cancellable, provider_load_cb,
+                                  g_steal_pointer (&task));
+}
+
+static void
+provider_load_cb (GObject      *source_object,
+                  GAsyncResult *result,
+                  gpointer      user_data)
+{
+  g_autoptr(GTask) task = G_TASK (user_data);
+  EpgService *self = EPG_SERVICE (g_task_get_source_object (task));
+  EpgProviderLoader *loader = EPG_PROVIDER_LOADER (source_object);
+  g_autoptr(GPtrArray) providers = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  providers = epg_provider_loader_load_finish (loader, result, &local_error);
+  if (providers == NULL)
+    {
+      g_warning ("%s: Failed to load external providers: %s",
+                 G_STRFUNC, local_error->message);
+      g_clear_error (&local_error);
+    }
+  else
+    {
+      EpgProvider *provider = NULL;
+      guint i;
```

Comment:
Nitpick: `gsize` for array iteration please

Reply:
I think it is correct to match the type of `GPtrArray.len`:

```c
struct GPtrArray {
  gpointer *pdata;
  guint	    len;
};
```

Reply:
(If I could go back in time and change `GPtrArray.len` to be a `gsize`, I would. Then we could legitimately address every byte of address space in a `GPtrArray` and the world would somehow be a better place.)

Reply:
And we could address more than 80 GiB of instances of EpgProvider implementations :-)